ports/raspberrypi/common-hal/busio/UART.c

```diff
@@ -264,106 +201,51 @@ size_t common_hal_busio_uart_read(busio_uart_obj_t *self, uint8_t *data, size_t
         mp_raise_ValueError(translate("No RX pin"));
     }
 
-    // This assignment is only here because the usart_async routines take a *const argument.
-    // struct usart_async_descriptor * const usart_desc_p = (struct usart_async_descriptor * const) &self->usart_desc;
-
     if (len == 0) {
         // Nothing to read.
         return 0;
     }
 
-    // struct io_descriptor *io;
-    // usart_async_get_io_descriptor(usart_desc_p, &io);
-
     size_t total_read = 0;
-    // uint64_t start_ticks = supervisor_ticks_ms64();
-
-    // // Busy-wait until timeout or until we've read enough chars.
-    // while (supervisor_ticks_ms64() - start_ticks <= self->timeout_ms) {
-    //     // Read as many chars as we can right now, up to len.
-    //     size_t num_read = io_read(io, data, len);
-
-    //     // Advance pointer in data buffer, and decrease how many chars left to read.
-    //     data += num_read;
-    //     len -= num_read;
-    //     total_read += num_read;
-    //     if (len == 0) {
-    //         // Don't need to read any more: data buf is full.
-    //         break;
-    //     }
-    //     if (num_read > 0) {
-    //         // Reset the timeout on every character read.
-    //         start_ticks = supervisor_ticks_ms64();
-    //     }
-    //     RUN_BACKGROUND_TASKS;
-    //     // Allow user to break out of a timeout with a KeyboardInterrupt.
-    //     if (mp_hal_is_interrupted()) {
-    //         break;
-    //     }
-    //     // If we are zero timeout, make sure we don't loop again (in the event
-    //     // we read in under 1ms)
-    //     if (self->timeout_ms == 0) {
-    //         break;
-    //     }
-    // }
-
-    // if (total_read == 0) {
-    //     *errcode = EAGAIN;
-    //     return MP_STREAM_ERROR;
-    // }
-
-    return total_read;
-}
-
-// Write characters.
-size_t common_hal_busio_uart_write(busio_uart_obj_t *self, const uint8_t *data, size_t len, int *errcode) {
-    if (self->tx_pin == NO_PIN) {
-        mp_raise_ValueError(translate("No TX pin"));
+    uint64_t start_ticks = supervisor_ticks_ms64();
+
+    // Busy-wait for all bytes received or timeout
+    while (len > 0 && (supervisor_ticks_ms64() - start_ticks < self->timeout_ms)) {
+        // Reset the timeout on every character read.
+        if (ringbuf_num_filled(&ringbuf[self->uart_id])) {
+            // Prevent conflict with uart irq.
+            irq_set_enabled(self->uart_irq_id, false);
+            // Copy as much received data as available, up to len bytes.
+            size_t num_read = ringbuf_get_n(&ringbuf[self->uart_id], data, len);
```

Comment:
There is a corner case here where the ringbuf and the FIFO could be filled before we read(). This code will copy from the ringbuf but not from the FIFO. Furthermore, if the FIFO is filled but not pulled from then we may end up with an infinite interrupt. Perhaps see what happens when a lot of data is sent but not read.